extmod/uasyncio/stream.py

```diff
@@ -30,6 +30,14 @@ def close(self):
         yield core._io_queue.queue_read(self.s)
         return self.s.read(n)
 
+    async def readexactly(self, n):
+        l = b""
+        while len(l) < n:
+            yield core._io_queue.queue_read(self.s)
+            l2 = self.s.read(n - len(l))
```

Comment:
If read hits an EOF then this loop will loop forever.  In CPython it will raise an `IncompleteReadError`.  In MicroPython I'm not sure what the best thing is to do in such an case.  It must at least terminate, and it would make sense to distinguish a short write with an exception, and that exception could be `EOFError` (which wouldn't cost anything extra since `EOFError` already exists as a built-in exception).

But, it would in some cases be important to retrieve the partially-read data.  `IncompleteReadError.partial` is used for this case.  So probably we need to implement that exception class and match the CPython behaviour for EOF...

Also, this code might be smaller/more efficient if the while is `while n` and `n` is decreased by `len(l2)` on each iteration (bytecode size can be checked by running `micropython -v -v -v`).

Reply:
I have updated the code with your suggestions with the option of raising EOFError.